staging/vhost-device-gpu/src/vhu_gpu.rs

```diff
@@ -384,164 +386,81 @@ impl VhostUserGpuBackend {
-                    ring_idx = ctrl_hdr.ring_idx;
-
-                    let fence = RutabagaFence {
-                        flags,
-                        fence_id,
-                        ctx_id,
-                        ring_idx,
-                    };
-                    gpu_response = match virtio_gpu.create_fence(fence) {
-                        Ok(_) => gpu_response,
-                        Err(fence_resp) => {
-                            warn!("create_fence {} -> {:?}", fence_id, fence_resp);
-                            fence_resp
-                        }
-                    };
-                }
-            }
-
-            // Prepare the response now, even if it is going to wait until
-            // fence is complete.
-
-            //edit reader definition later to something better signifying response wrte
-            match gpu_response.encode(flags, fence_id, ctx_id, ring_idx, &reader, desc_hdraddr) {
-                Ok(l) => len = l,
-                Err(e) => debug!("ctrl queue response encode error: {:?}", e),
-            }
-
-            if flags & VIRTIO_GPU_FLAG_FENCE != 0 {
-                let ring = match flags & VIRTIO_GPU_FLAG_INFO_RING_IDX {
-                    0 => VirtioGpuRing::Global,
-                    _ => VirtioGpuRing::ContextSpecific { ctx_id, ring_idx },
-                };
-
-                add_to_queue =
-                    virtio_gpu.process_fence(ring, fence_id, desc_chain.head_index(), len);
-            }
-            if add_to_queue {
-                used_len += desc_hdr.len();
-                if vring.add_used(desc_chain.head_index(), used_len).is_err() {
-                    log::error!("Couldn't return used descriptors to the ring");
-                }
-                used_any = true;
-            }
-
-            // // Keep track of bytes that will be written in the VQ.
-            // let mut used_len = 0;
+        }
 
-            // // Response Header descriptor.
-            // let desc_response = descriptors[1];
-            // if !desc_response.is_write_only() {
-            //     return Err(Error::UnexpectedReadableDescriptor(1).into());
-            // }
-            // let gpu_cmd_type = GpuCommandType::try_from(request.gpu_type).map_err(Error::from)?;
-            // //let resp = self.process_gpu_command(gpu_cmd_type);
+        // Prepare the response now, even if it is going to wait until
+        // fence is complete.
+        let response_len = response
+            .encode(hdr.flags, hdr.fence_id, hdr.ctx_id, hdr.ring_idx, writer)
+            .map_err(Error::GpuResponseEncode)?;
+
+        let mut add_to_queue = true;
+        if hdr.flags & VIRTIO_GPU_FLAG_FENCE != 0 {
+            let ring = match hdr.flags & VIRTIO_GPU_FLAG_INFO_RING_IDX {
+                0 => VirtioGpuRing::Global,
+                _ => VirtioGpuRing::ContextSpecific {
+                    ctx_id: hdr.ctx_id,
+                    ring_idx: hdr.ring_idx,
```

Comment:
hummm, storing hdr.flags in a mutable variable might be more appropriate. don't you think so?

Reply:
Oops, I think I changed the behavior with this (it may do something different when the command doesn't have a fence flag.), I'll change it to the way crossvm does it.